### ISSUE TITLE
Update Tasks

### ISSUE THREAD

**Issue by MaxTwentythree** [veliovgroup/josk#7]
Hi,

I'd like to update a task that already exists.
As this doesn't seem to be possible, my current workaround is to remove a task if it already exists and recreate it with the new values.

Firstly I wonder why it's necessary to append "setTimeout" / "setInterval" to the task uid?
Wouldn't it be better to be able to access a task with the uid it was actually created with?
Alternatively it would be nice to maybe have something like an exposed generateUid method so that I don't have to append setTimeout manually (for the case this logic might change later)..

Secondly removing a task and recreating it within the same Meteor method doesn't work.
(I think it's what you mean with "Should be called in a separate event loop").
I guess I could trigger the same Meteor method after deleting the task, but I'm not sure if this would even work and it feels kinda error-prone..

It does work if I do it manually, but I'm pretty sure that's bad practice:

```
const existingJob = job.collection.findOne({uid: `${taskId}setTimeout`});
if (existingJob) job.collection.remove({uid: `${taskId}setTimeout`});
...
job.setTimeout(task, diff, taskId)
```
Is there a proposed solution to achieving this?

**Comment by dr-dimitru** [comment]
Hello @MaxTwentythree ,

Updates would work behind the scenes. Just create a new task with same task id "uid", and it [would get updated](https://github.com/veliovgroup/josk/blob/master/index.js#L262).

> Firstly I wonder why it's necessary to append "setTimeout" / "setInterval" to the task uid?

No, package does it for you. The reason — be able to create Interval and Timeout tasks with same uid. To be honest — I don't see in the docs suggesting `uid` to be in the form of `${taskId}setTimeout`

> Secondly removing a task and recreating it within the same Meteor method doesn't work.

I'd like to see code snippet where yu remove the task and that doesn't work.

> (I think it's what you mean with "Should be called in a separate event loop").

Here's example:
```js
const timer = job.setTimeout(func, 34789, 'unique-taskid');
someAsyncCall(() => {
  job.clearTimeout(timer);
});
```

Another example:
```js
const timer = job.setTimeout(func, 34789, 'unique-taskid');
process.nextTick(() => {
  job.clearTimeout(timer);
});
```

> Is there a proposed solution to achieving this?

What was your original question? Regarding updating existing task:

```js
// CREATE NEW TASK
const timer = job.setTimeout(func, 34789, 'unique-taskid');

// UPDATE EXISTING TASK
const timer = job.setTimeout(otherFunc, 512, 'unique-taskid'); // <-- same 'unique-taskid'
```

**Comment by MaxTwentythree** [comment]
Hi @dr-dimitru
Ah, super, that's exactly what I need! 
Apparently I did not try the obvious one :D
Those snippets might also be nice for the documentation..
Thanks a lot for your help and the great package!

**Comment by dr-dimitru** [comment]
@MaxTwentythree I'm glad that was solved so quickly 😅
Thank you for documentation suggestion, let's keep it open until docs is updated

**Comment by MaxTwentythree** [comment]
Hi @dr-dimitru,
I now tried the update functionality but it doesn't seem to work :( .. maybe I did something wrong
Here's my method

```
  async [methods.addTask](
    nodeId: string,
    endDate: Date
  ): Promise<void> {
    check(nodeId, String);
    check(endDate, Date);

    const now = new Date();
    const diff = endDate.getTime() - now.getTime();

    const db = Meteor.users.rawDatabase();
    const job = new JoSk({db});
    const taskId = `task ${nodeId}`;

    const bound = Meteor.bindEnvironment((callback: () => void) => {
      callback();
    });

    const task = function finishFeedbackPhaseTask(ready: () => void): void {
      bound(() => {
        Meteor.call(methods.someTask, nodeId, () => {
          ready();
        });
      });
    };

    job.setTimeout(task, diff, taskId);
  }
```

When I call this method twice with a different enddate (but the same nodeId, which means same taskId) the task should update its executeAt prop, right?

**Comment by dr-dimitru** [comment]
@MaxTwentythree 

1. Yes, `delay` and `executeAt` properties [should get updated](https://github.com/veliovgroup/josk/blob/master/index.js#L266)
2. You should initiate `JoSk` only once, move `const job = new JoSk({db});` out of the `addTaskMethod` function to global scope. Ideally you should initiate it in its own file and `export { job }`, then `import` where used

Let me know if it helps

**Comment by MaxTwentythree** [comment]
Ah, I see, 2) makes a lot of sense ;)
I fixed this, but and initiate JoSk now only once.
But it didn't fix the update unfortunatelly.
I also tried the josk onError callback, but I don't get errors for the not updated tasks there..
It's not a big deal as I have a clearJob method anyways, which I now run before creating new tasks and it works like this.
Still curious why it doesn't work without it though..

**Comment by dr-dimitru** [comment]
@MaxTwentythree will add it to our test suit, let's see if we can replicate it. Let's keep it open for now, thank you for reporting

**Comment by MaxTwentythree** [comment]
Thanks a lot @dr-dimitru !

**Comment by dr-dimitru** [comment]
@MaxTwentythree it's been a while, no changes to the codebase were necessary. Please, see [`override` tests](https://github.com/veliovgroup/josk/blob/2.4.1/test/npm.js#L179). The main difference between this and your examples, — `.find()` called in the separate event-loop from creating and updating tsk

https://github.com/veliovgroup/josk/blob/464a7adc4f5be4d0da3069d1948154ef3c54aa45/test/npm.js#L184-L206


Feel free to reopen it in case if the issue is still persists on your end.
